controllers/topology/desired_state_test.go

```diff
@@ -390,6 +391,233 @@ func TestComputeControlPlane(t *testing.T) {
 			obj:         obj,
 		})
 	})
+	t.Run("Should choose the correct version for control plane", func(t *testing.T) {
+		// Note: in all of the following tests we are setting it up so that there are not machine deployments.
+		// A more extensive list of scenarios is tested in TestComputeControlPlaneVersion.
+		tests := []struct {
+			name                string
+			currentControlPlane *unstructured.Unstructured
+			topologyVersion     string
+			expectedVersion     string
+		}{
+			{
+				name:                "use cluster.spec.topology.version if creating a new control plane",
+				currentControlPlane: nil,
+				topologyVersion:     "v1.2.3",
+				expectedVersion:     "v1.2.3",
+			},
+			{
+				name: "use controlplane.spec.version if the control plane's spec.version is not equal to status.version",
+				currentControlPlane: testtypes.NewControlPlaneBuilder("test1", "cp1").
+					WithSpecFields(map[string]interface{}{
+						"spec.version": "v1.2.2",
+					}).
+					WithStatusFields(map[string]interface{}{
+						"status.version": "v1.2.1",
+					}).
+					Build(),
+				topologyVersion: "v1.2.3",
+				expectedVersion: "v1.2.2",
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.name, func(t *testing.T) {
+				g := NewWithT(t)
+
+				// Current cluster objects for the test scenario.
+				clusterWithControlPlaneRef := cluster.DeepCopy()
+				clusterWithControlPlaneRef.Spec.ControlPlaneRef = fakeRef1
+				clusterWithControlPlaneRef.Spec.Topology.Version = tt.topologyVersion
+
+				blueprint := &scope.ClusterBlueprint{
+					Topology:     clusterWithControlPlaneRef.Spec.Topology,
+					ClusterClass: clusterClass,
+					ControlPlane: &scope.ControlPlaneBlueprint{
+						Template: controlPlaneTemplate,
+					},
+				}
+
+				// Aggregating current cluster objects into ClusterState (simulating getCurrentState).
+				s := scope.New(clusterWithControlPlaneRef)
+				s.Blueprint = blueprint
+				s.Current.ControlPlane = &scope.ControlPlaneState{
+					Object: tt.currentControlPlane,
+				}
+
+				obj, err := computeControlPlane(ctx, s, nil)
+				g.Expect(err).NotTo(HaveOccurred())
+				g.Expect(obj).NotTo(BeNil())
+				assertNestedField(g, obj, tt.expectedVersion, contract.ControlPlane().Version().Path()...)
+			})
+		}
+	})
+}
+
+func TestComputeControlPlaneVersion(t *testing.T) {
+	// Note: the version used by the machine deployments does
+	// not affect how we determining the control plane version.
+	// We only want to know if the machine deployments are stable.
+	//
+	// A machine deployment is considere stable if all the following are true:
+	// - md.spec.replicas == md.status.replicas
+	// - md.spec.replicas == md.status.updatedReplicas
+	// - md.spec.replicas == md.status.readyReplicas
+	// - md.Generation < md.status.observedGeneration
+	//
+	// A machine deployment is considered upgrading if any of the above conditions
```

Comment:
Similar to comment on the isUpgrading method - upgrade has a specific meaning so maybe there's a better term here? Something simpler like updating might work, or even 'stable' defined as the above state. IsStable() etc.